### PR TITLE
Add support for limit pushdown through union

### LINKED ISSUE #14894: limit pushdown through union
add support to allow limit pushdown through union node

### PR DESCRIPTION
**Pull request by fgwang7w**
```
== RELEASE NOTES ==
General Changes
* Add support for limit pushdown through union
```

resolve #14894

### CODE REVIEW

**Comment by kaikalur** [comment]
Will be nice if this applies to ORDER BY LIMIT as well.

LGTM

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPushLimitThroughUnion.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+    @Test
+    public void doesNotFire()
+    {
+        tester().assertThat(new PushLimitThroughUnion())
+                .on(p -> {
+                    VariableReferenceExpression a = p.variable("a");
+                    VariableReferenceExpression b = p.variable("b");
+                    VariableReferenceExpression c = p.variable("c");
+                    return p.limit(1,
```
Comment (rongrong) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPushLimitThroughUnion.java:62]: So, theoretically, this won't fire if the limit here is > 1 either right? Maybe also test that?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushLimitThroughUnion.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(LimitNode parent, Captures captures, Context context)
+    {
+        UnionNode unionNode = captures.get(CHILD);
+        ImmutableList.Builder<PlanNode> builder = ImmutableList.builder();
+        boolean shouldApply = false;
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushLimitThroughUnion.java:71]: I prefer a separate check before building the list. Something like
```
if (unionNode.getSources().stream().allMatch(source -> isAtMost(source, context.getLookup(), parent.getCount()))) {
    return Result.empty();
}
// build new plan node
```
It's slightly more efficient when the rule should not fire, which should happen more often.

Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushLimitThroughUnion.java:71]: yes this check makes sense, will incorporate it.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPushLimitThroughUnion.java (hunk 1 above)
Comment (fgwang7w) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPushLimitThroughUnion.java:62]: sure I add another test to cover limit2 case

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushLimitThroughUnion.java (hunk 2 above)
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushLimitThroughUnion.java:74]: You don't need to check `shouldApply` anymore after the check. It should always apply if the code gets here.

**Pull request merged**